test/validation/api/timer/timer.c

```diff
@@ -796,67 +798,73 @@ static void timer_single_shot(odp_queue_type_t queue_type, odp_timer_tick_type_t
-
-	start_param.tick_type = tick_type;
-	start_param.tick      = tick;
-	start_param.tmo_ev    = ev;
-
-	ret = odp_timer_start(timer, &start_param);
-	CU_ASSERT_FATAL(ret == ODP_TIMER_SUCCESS);
-
-	if (restart) {
-		tick = odp_timer_ns_to_tick(tp, tmo_ns);
+	for (i = 0; i < rounds; i++) {
+		tick = odp_timer_ns_to_tick(tp, nsec);
 		if (tick_type == ODP_TIMER_TICK_ABS)
 			tick += odp_timer_current_tick(tp);
 
-		start_param.tick   = tick;
-		start_param.tmo_ev = ODP_EVENT_INVALID;
+		start_param.tick_type = tick_type;
+		start_param.tick      = tick;
+		start_param.tmo_ev    = ev;
 
-		ret = odp_timer_restart(timer, &start_param);
+		ret = odp_timer_start(timer, &start_param);
 		CU_ASSERT_FATAL(ret == ODP_TIMER_SUCCESS);
-	}
 
-	ev = ODP_EVENT_INVALID;
+		if (restart) {
+			tick = odp_timer_ns_to_tick(tp, tmo_ns);
+			if (tick_type == ODP_TIMER_TICK_ABS)
+				tick += odp_timer_current_tick(tp);
 
-	if (cancel) {
-		ret = odp_timer_cancel(timer, &ev);
-		CU_ASSERT(ret == 0);
+			start_param.tick   = tick;
+			start_param.tmo_ev = ODP_EVENT_INVALID;
 
-		if (ret == 0)
-			CU_ASSERT(ev != ODP_EVENT_INVALID);
-	} else {
-		uint64_t diff_ns;
+			ret = odp_timer_restart(timer, &start_param);
+			CU_ASSERT_FATAL(ret == ODP_TIMER_SUCCESS);
+		}
 
-		t1 = odp_time_global();
-		ev = wait_event(queue_type, queue, t1, 10 * tmo_ns);
-		t2 = odp_time_global();
-		diff_ns = odp_time_diff_ns(t2, t1);
+		ev = ODP_EVENT_INVALID;
 
-		CU_ASSERT(ev != ODP_EVENT_INVALID);
-		CU_ASSERT(diff_ns < 2 * tmo_ns);
-		CU_ASSERT((double)diff_ns > 0.5 * tmo_ns);
-	}
+		if (cancel) {
+			ret = odp_timer_cancel(timer, &ev);
+			CU_ASSERT(ret == 0);
 
-	if (ev != ODP_EVENT_INVALID) {
-		CU_ASSERT_FATAL(odp_event_type(ev) == ODP_EVENT_TIMEOUT);
-		tmo = odp_timeout_from_event(ev);
-		CU_ASSERT(odp_timeout_user_ptr(tmo) == USER_PTR);
-		CU_ASSERT(odp_timeout_timer(tmo) == timer);
+			if (ret == 0)
+				CU_ASSERT(ev != ODP_EVENT_INVALID);
+		} else {
+			uint64_t diff_ns;
 
-		if (!cancel) {
-			if (tick_type == ODP_TIMER_TICK_ABS) {
-				/* CU_ASSERT needs these extra brackets */
-				CU_ASSERT(odp_timeout_tick(tmo) == tick);
-			} else {
-				CU_ASSERT(odp_timeout_tick(tmo) > tick);
+			t1 = odp_time_global();
+			ev = wait_event(queue_type, queue, t1, 10 * tmo_ns);
+			t2 = odp_time_global();
+			diff_ns = odp_time_diff_ns(t2, t1);
+
+			CU_ASSERT(ev != ODP_EVENT_INVALID);
+			CU_ASSERT(diff_ns < 2 * tmo_ns);
+			CU_ASSERT((double)diff_ns > 0.5 * tmo_ns);
+		}
+
+		if (ev != ODP_EVENT_INVALID) {
+			CU_ASSERT_FATAL(odp_event_type(ev) == ODP_EVENT_TIMEOUT);
+			tmo = odp_timeout_from_event(ev);
+			CU_ASSERT(odp_timeout_user_ptr(tmo) == USER_PTR);
+			CU_ASSERT(odp_timeout_timer(tmo) == timer);
+
+			if (!cancel) {
+				if (tick_type == ODP_TIMER_TICK_ABS) {
+					/* CU_ASSERT needs these extra brackets */
+					CU_ASSERT(odp_timeout_tick(tmo) == tick);
+				} else {
+					CU_ASSERT(odp_timeout_tick(tmo) > tick);
+				}
 			}
+		} else {
+			ODPH_DBG("Event missing\n");
```

Comment:
On the other hand, I guess earlier asserts have already been triggered at this point and no need to output anything here.

Reply:
Yes, there is an assert on line 841 for catching this:
CU_ASSERT(ev != ODP_EVENT_INVALID);

Since asserts are reported only after test case has stopped, I added this extra print to help debugging.